packages/core/src/config/config.ts

```diff
@@ -200,70 +55,89 @@ export type Network = {
+>;
+
+type GetContract<networks = unknown, contract = unknown> = contract extends {
+  abi: infer abi extends Abi;
+}
+  ? // 1. Contract has a valid abi
+    ContractConfig<networks, contract, abi>
+  : // 2. Contract has an invalid abi
+    ContractConfig<networks, contract, Abi>;
+
+type ContractsConfig<networks, contracts> = {} extends contracts
+  ? // contracts empty, return empty
+    {}
+  : {
+      [name in keyof contracts]: GetContract<networks, contracts[name]>;
+    };
+
+type NetworksConfig<networks> = {} extends networks
+  ? {}
+  : {
+      [networkName in keyof networks]: NetworkConfig;
+    };
+
+export const createConfig = <const networks, const contracts>(config: {
+  // TODO: add jsdoc to these properties.
+  networks: NetworksConfig<networks>;
+  contracts: ContractsConfig<networks, contracts>;
+  database?: DatabaseConfig;
+  options?: OptionConfig;
+}): CreateConfigReturnType<networks, contracts> =>
+  config as CreateConfigReturnType<networks, contracts>;
+
 export type Config = {
-  /** Database to use for storing raw & indexed data. Default: `"postgres"` if `DATABASE_URL` env var is present, otherwise `"sqlite"`. */
-  database?: Database;
-  networks: Record<string, Network>;
-  /** List of contracts to sync & index events from. Contracts defined here will be present in `context.contracts`. */
-  contracts: Record<string, Contract<string, Abi, string, undefined>>;
-  /** Configuration for Ponder internals. */
-  options?: Option;
+  networks: { [name: string]: NetworkConfig };
+  contracts: { [name: string]: GetContract };
+  database?: DatabaseConfig;
+  options?: OptionConfig;
 };
 
-/**
- * Create a Ponder config.
- *
- * - Docs: https://ponder.sh/docs/api-reference/config
- *
- * @example
- * export default createConfig({
- *   networks: {
- *     mainnet: { chainId: 1, transport: http(process.env.PONDER_RPC_URL_1) },
- *   },
- *   contracts: {
- *     ERC20: {
- *       abi: ERC20Abi,
- *       network: "mainnet",
- *       address: "0x123...",
- *       startBlock: 1500,
- *     },
- *   },
- * })
- */
-export const createConfig = <
-  const TNetworks extends Record<string, Network>,
-  const TContracts extends {
-    [ContractName in keyof TContracts]: Contract<
-      keyof TNetworks & string,
-      TContracts[ContractName]["abi"],
-      TContracts[ContractName]["filter"] extends {
-        event: infer _event extends string;
-      }
-        ? _event
-        : string,
-      TContracts[ContractName] extends {
-        factory: {
-          event: infer _event extends AbiEvent;
-        };
-      }
-        ? _event
-        : undefined
-    >;
-  },
->(config: {
-  database?: Database;
-  networks: TNetworks;
-  contracts: TContracts &
-    Record<string, Contract<string, Abi, string, AbiEvent>>;
-  options?: Option;
-}): {
-  database?: Database;
-  networks: TNetworks;
-  contracts: TContracts;
-  options?: Option;
-} => config;
+export type CreateConfigReturnType<networks, contracts> = {
```

Comment:
What's the purpose of this type?

Reply:
It enforces maximum strictness. The type "Config" is not as strict because we want the nice language server features when hovering over a property.